builder/config_reader.go

```diff
@@ -95,31 +96,53 @@ func ReadConfig(path string) (config Config, warnings []string, err error) {
 	return config, warnings, nil
 }
 
-func getWarningsForObsoleteFields(reader io.Reader) ([]string, error) {
+func getWarningsForObsoleteFields(reader io.Reader) ([]string, int, error) {
 	var warnings []string
 
 	var obsoleteConfig = struct {
 		Groups []interface{}
 	}{}
 
 	if _, err := toml.DecodeReader(reader, &obsoleteConfig); err != nil {
-		return nil, err
+		return nil, 0, err
 	}
 
 	if len(obsoleteConfig.Groups) > 0 {
 		warnings = append(warnings, fmt.Sprintf("%s field is obsolete in favor of %s", style.Symbol("groups"), style.Symbol("order")))
 	}
 
-	return warnings, nil
+	return warnings, len(obsoleteConfig.Groups), nil
 }
 
 // parseConfig reads a builder configuration from reader and resolves relative buildpack paths using `relativeToDir`
-func parseConfig(reader io.Reader, relativeToDir string) (Config, error) {
-	var builderConfig Config
-	if _, err := toml.DecodeReader(reader, &builderConfig); err != nil {
+func parseConfig(reader io.Reader, relativeToDir, path string, obsoleteConfigs int) (Config, error) {
+	builderConfig := Config{}
+	fmt.Println(obsoleteConfigs)
+	tomlMetadata, err := toml.DecodeReader(reader, &builderConfig)
+	if err != nil {
 		return Config{}, errors.Wrap(err, "decoding toml contents")
 	}
 
+	undecodedKeys := tomlMetadata.Undecoded()
+	if len(undecodedKeys) > 0 {
+		errorKeys := config.ParseUndecodedKeys(undecodedKeys)
+
+		pluralizedElement := "element"
+		if len(errorKeys) > 1 {
+			pluralizedElement += "s"
+		}
+
+		if obsoleteConfigs > 0 {
+			return builderConfig, nil
+		}
```

Comment:
If there are other invalid elements in addition to `groups` in the builder.toml, this logic would result in us not failing as we intend.

Reply:
Actually, we've open https://github.com/buildpacks/pack/issues/538 to remove the warning completely and just error when `groups` (or any other invalid config) is provided. Given that, I think we can accept this implementation in this pull request, and address obsolete configs in the other issue.